caas/kubernetes/provider/metadata.go

```diff
@@ -81,83 +98,88 @@ func (k *kubernetesClient) GetClusterMetadata(storageClass string) (*caas.Cluste
 		if err != nil && !k8serrors.IsNotFound(err) {
 			return nil, errors.Trace(err)
 		}
-		if err == nil {
-			result.NominatedStorageClass = &caas.StorageProvisioner{
-				Name:        sc.Name,
-				Provisioner: sc.Provisioner,
-				Parameters:  sc.Parameters,
-			}
-			if sc.ReclaimPolicy != nil {
-				result.NominatedStorageClass.ReclaimPolicy = string(*sc.ReclaimPolicy)
-			}
+		if err == nil && sc != nil {
+			logger.Debugf("Use %q for nominated storage class", sc.Name)
+			result.NominatedStorageClass = caasStorageProvisioner(*sc)
 		}
 	}
 
 	// We may have the workload storage but still need to look for operator storage.
-	preferredOperatorStorage, havePreferredOperatorStorage := jujuPreferredOperatorStorage[result.Cloud]
 	storageClasses, err := k.client().StorageV1().StorageClasses().List(v1.ListOptions{})
 	if err != nil {
 		return nil, errors.Annotate(err, "listing storage classes")
 	}
 
-	var (
-		possibleWorkloadStorage []storage.StorageClass
-		possibleOperatorStorage []*caas.StorageProvisioner
-		defaultOperatorStorage  *caas.StorageProvisioner
-	)
-	for _, sc := range storageClasses.Items {
-		if havePreferredOperatorStorage {
-			maybeOperatorStorage := &caas.StorageProvisioner{
-				Name:        sc.Name,
-				Provisioner: sc.Provisioner,
-				Parameters:  sc.Parameters,
-			}
-			if sc.ReclaimPolicy != nil {
-				maybeOperatorStorage.ReclaimPolicy = string(*sc.ReclaimPolicy)
+	var possibleWorkloadStorage, possibleOperatorStorage []*caas.StorageProvisioner
+	preferredOperatorStorage, hasPreferredOperatorStorage := jujuPreferredOperatorStorage[result.Cloud]
+
+	pickOperatorSC := func(sc storage.StorageClass, caasSC *caas.StorageProvisioner) {
```

Comment:
Can these pickXXXX functions be moved out?

Reply:
The reason why they are not moved out of the method was because the params could be getting too long. They are sharing lots of local vars